lib/sshutils/sftp/sftp.go

```diff
@@ -423,6 +425,49 @@ func (c *Config) transferFile(ctx context.Context, dstPath, srcPath string, srcF
 	return nil
 }
 
+// assertStreamsType checks if reader or writer implements correct interface to utilize concurrent SFTP streams.
+func assertStreamsType(reader io.Reader, writer io.Writer) error {
+	_, okReader := reader.(io.WriterTo)
+	_, okWriter := writer.(io.ReaderFrom)
+
+	if !okWriter && !okReader {
```

Comment:
Should this be "or" or is it ok for just one of them to implement it?
```suggestion
	if !okWriter || !okReader {
```

Reply:
Just one needs to implement it. If we copy a file from local to remote, then `remote`(writer) needs to implement `io.ReadFrom` interface. The local side won't, as we need to call `Read/Write` to get the progress bar. Copying from remote to local has reverted conditions. I know this is super confusing, but I've no better idea how to assert this. Go sftp doesn't enforce any type, just do a few runtime checks on copy, and if "by mistake" all interfaces are implemented, it will allow concurrent copy.